### ISSUE TITLE
@typing.dataclass_transform and custom __hash__ in subclass

### ISSUE THREAD

**Issue by AlexanderPodorov** [microsoft/pyright#5100]
**Describe the bug**
Example code throws (using `pylance` **strict** mode):
```python
"__hash__" overrides symbol of same name in class "A"
  Type cannot be assigned to type "None"
```

**To Reproduce**
See example code.

**Expected behavior**
No warnings thrown.

**Screenshots or Code**
```python
from typing import dataclass_transform


@dataclass_transform()
class Meta(type):
    pass


class A(metaclass=Meta):
    pass


class B(A):

    def __hash__(self) -> int:
        return 0
```

**VS Code extension or command-line**
pylance version 2023.5.20

**Additional context**
[See here](https://github.com/pydantic/pydantic/issues/5726)

**Comment by erictraut** [comment]
A stdlib `dataclass` has a synthesized attribute `__hash__` that is set to `None` if the following conditions exist: 1) the dataclass isn't frozen, 2) the class doesn't define an explicit `__hash__` method, and 3) the `hash` (or `unsafe_hash`) parameter hasn't been set to True. 

I don't know if pydantic has the same behavior as stdlib's `dataclass`, but the `dataclass_transform` mechanism implies all of the stdlib `dataclass` behaviors from the perspective of a type checker.

I presume that you've enabled the strict-mode `reportIncompatibleVariableOverride` rule. This check is designed to detect cases where you are overriding a base class variable of the same name in an incompatible manner. Since the base class (`A`) implicitly implements a variable called `__hash__` which has the type `None`, this error is technical correct. However, I can see how it would be confusing in this case.

I've updated pyright to exempt the dataclass `__hash__` symbol from override checks. This will be included in the next release of pyright.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.308, which I just published. It will also be included in a future release of pylance.
